cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -1479,15 +1479,22 @@ private module Cached {
     // Reverse flow: data that flows from the definition node back into the indirection returned
     // by a function. This allows data to flow 'in' through references returned by a modeled
     // function such as `operator[]`.
-    exists(Operand address, int indirectionIndex |
-      nodeHasOperand(nodeTo.(IndirectReturnOutNode), address, indirectionIndex)
+    or
+    exists(
+      DataFlowFunction f, CallInstruction call, FunctionInput inModel, FunctionOutput outModel,
+      int indirectionIndex, IndirectReturnOutNode outNode
     |
-      exists(StoreInstruction store |
-        nodeHasInstruction(nodeFrom, store, indirectionIndex - 1) and
-        store.getDestinationAddressOperand() = address
-      )
-      or
-      Ssa::outNodeHasAddressAndIndex(nodeFrom, address, indirectionIndex)
+      inModel.isReturnValueDeref() and
+      outModel.isQualifierObject() and
+      f.hasDataFlow(inModel, outModel) and
+      call.getStaticCallTarget() = f and
+      // nodeFrom is a post-update node whose pre-update node is an out node
+      outNode = nodeFrom.(PostUpdateNode).getPreUpdateNode() and
+      outNode.getCallInstruction() = call and
+      outNode.getIndirectionIndex() = indirectionIndex and
```

Comment:
I'm not sure this join orders nicely as written. If the first conjunct is joined first we get a tuple explosion on the order of "the number of indirections" for the out node, and if the second conjunct is chosen first we get a tuple explosion on the order of "number of calls in the database".

I think you want to factor out a three-column predicate that joins on both `call` and `indirectionIndex`, so that you can do a nice join with this conjunct:
```ql
hasOperandAndIndex(nodeTo.(PostUpdateNode).getPreUpdateNode(), call.getThisArgumentOperand(), indirectionIndex)
```